### PR TITLE
NO-TICKET: fold documentation

### PR DESCRIPTION
**Pull request by cateland**
What this PR does / why we need it:
Documentation for the `fold` function
Which issue(s) this PR fixes (optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged): Fixes #

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -103,6 +103,40 @@ export const append = <E, A, B>(
   );
 };
 
+/**
+ * Fold is the only way to attempt to safely extract the data,
+ * because it force you to consider the four variation of the dataway state.
```
Comment (graffardi) [inline src/main.ts]: forces à la place de force je pense

File: src/main.ts [hunk 2]
```diff
@@ -103,6 +103,40 @@ export const append = <E, A, B>(
...
+ * The following example demostrate how usefull this is when trying to render a
```
Comment (graffardi) [inline src/main.ts]: demonstrates, useful

File: src/main.ts (hunk 1 above)
Comment (guillaumewuip) [inline src/main.ts]:
```suggestion
 * because it force you to consider the four variations of the dataway state.
```

File: src/main.ts (hunk 2 above)
Comment (guillaumewuip) [inline src/main.ts]:
```suggestion
 * The following example demonstrate how useful this is when trying to render a
```

File: src/main.ts [hunk 3]
```diff
@@ -103,6 +103,40 @@ export const append = <E, A, B>(
...
+ * search result, we may want to notify the user why no result show up on his screen
```
Comment (guillaumewuip) [inline src/main.ts]:
```suggestion
 * search result. We may want to notify the user why no result show up on his screen.
```

File: src/main.ts [hunk 4]
```diff
@@ -103,6 +103,40 @@ export const append = <E, A, B>(
...
+ * 1. The search could not have been initiated by the user (he didn't click on the search button),
+ *      a loading visual hint coulb be displayed
```
Comment (guillaumewuip) [inline src/main.ts]:
```suggestion
 *      a loading visual hint could be displayed
```

Comment (euphocat) [inline src/main.ts]:
```suggestion
 *      a loading visual hint could be displayed
```

File: src/main.ts (hunk 3 above)
Comment (graffardi) [inline src/main.ts]: showed up

File: src/main.ts [hunk 5]
```diff
@@ -103,6 +103,40 @@ export const append = <E, A, B>(
...
+ * 2. The search could be initiated but not done yet (search can take time),
+ *      an error message can be shown or a message to try his luck again.
+ * 3. The search could end up in error (timeout, connectivity issue or plain server crash)
```
Comment (guillaumewuip) [inline src/main.ts]:
```suggestion
 * 3. The search could end up in error (timeout, connectivity issue or plain server crash), 
 *      an error message can be shown or a message to try again.
```

File: src/main.ts [hunk 6]
```diff
@@ -103,6 +103,40 @@ export const append = <E, A, B>(
...
 
+/**
+ * Fold is the only way to attempt to safely extract the data,
+ * because it force you to consider the four variation of the dataway state.
+ * The following example demostrate how usefull this is when trying to render a
+ * search result, we may want to notify the user why no result show up on his screen
+ * 1. The search could not have been initiated by the user (he didn't click on the search button),
+ *      a loading visual hint coulb be displayed
+ * 2. The search could be initiated but not done yet (search can take time),
+ *      an error message can be shown or a message to try his luck again.
```
Comment (euphocat) [inline src/main.ts]:
```suggestion
 *      an error message can be shown or a message to try again.
```

File: src/main.ts (hunk 4 above)
Comment (guillaumewuip) [inline src/main.ts]:
```suggestion
```

File: src/main.ts (hunk 6 above)
Comment (guillaumewuip) [inline src/main.ts]:
```suggestion
 *      a loading visual hint could be displayed
```

File: src/main.ts [hunk 7]
```diff
@@ -103,6 +103,40 @@ export const append = <E, A, B>(
...
+ * 3. The search could end up in error (timeout, connectivity issue or plain server crash)
+ * 4. The search can success and we want to display the result to the user.
+ * 
+ * ```
+ * import { dataway, fold } from "dataway";
+ *
+ * const searchResult = dataway.of(["result1", "result2", "result3", "result4"]);
+ *
+ * fold(
+ *   () => "<span>Search for nice code snippet</span>",
```
Comment (guillaumewuip) [inline src/main.ts]:
```suggestion
 *   () => "<button onClick={...}>Search for nice code snippet</button>",
```

File: src/main.ts [hunk 8]
```diff
@@ -103,6 +103,40 @@ export const append = <E, A, B>(
...
 };
 
+/**
+ * Fold is the only way to attempt to safely extract the data,
+ * because it force you to consider the four variation of the dataway state.
+ * The following example demostrate how usefull this is when trying to render a
+ * search result, we may want to notify the user why no result show up on his screen
+ * 1. The search could not have been initiated by the user (he didn't click on the search button),
+ *      a loading visual hint coulb be displayed
+ * 2. The search could be initiated but not done yet (search can take time),
```
Comment (graffardi) [inline src/main.ts]: `notAsked` et le `loading` sont inversés je pense

```suggestion
 * 1. The search could be initiated but not done yet (he didn't click on the search button),
 *      nothing should be happening
 * 2. The search has been initiated by the user (search can take time),
 *      a loading visual hint could be displayed
 * 3. The search could end up in error (timeout, connectivity issue or plain server crash)
 * 4. The search can success and we want to display the result to the user.
```

Comment (guillaumewuip) [reply src/main.ts]: c'est fixé

**Pull request merged**